File: scala/src/main/scala/folio/codinginterview/presentation/PortfolioController.scala
package folio.codinginterview.presentation

import folio.codinginterview.application.usecase.portfolio.GetLatestPortfolioUsecase
import folio.codinginterview.application.usecase.portfolio.UpdatePortfolioItemInput
import folio.codinginterview.application.usecase.portfolio.UpdatePortfolioUsecase
import folio.codinginterview.application.usecase.portfolio.UpdatePortfolioUsecaseException
import folio.codinginterview.application.usecase.portfolio.UpdatePortfolioUsecaseInput
import folio.codinginterview.domain.StockSymbol
import folio.codinginterview.presentation.PresentationException.BadRequestException
import scala.concurrent.ExecutionContext
import scala.concurrent.Future

object PortfolioController {
  final case class PortfolioItemDto(symbol: String, rate: String)
  final case class GetOptimalPortfolioResponse(portfolios: Seq[PortfolioItemDto])
  final case class UpdateOptimalPortfolioRequest(portfolios: Seq[PortfolioItemDto])
}

final class PortfolioController(
    getLatestPortfolioUsecase: GetLatestPortfolioUsecase,
    updatePortfolioUsecase: UpdatePortfolioUsecase
)(using ec: ExecutionContext) {
  import PortfolioController.*

  def getOptimalPortfolio(): Future[GetOptimalPortfolioResponse] = {
    getLatestPortfolioUsecase.run().map { out =>
      GetOptimalPortfolioResponse(
        out.items.map(i => PortfolioItemDto(i.symbol.toString, i.rate.toString))
      )
    }
  }

  def updateOptimalPortfolio(req: UpdateOptimalPortfolioRequest): Future[Unit] = {
    val parsed: Either[String, Seq[UpdatePortfolioItemInput]] =
      req.portfolios.foldLeft[Either[String, Seq[UpdatePortfolioItemInput]]](Right(Seq.empty)) { case (acc, dto) =>
        for {
          xs <- acc
          sym <- StockSymbol
            .fromString(dto.symbol)
            .toRight(s"unknown symbol: ${dto.symbol}")
          rate <-
            try Right(BigDecimal(dto.rate))
            catch { case _: Throwable => Left(s"invalid rate: ${dto.rate}") }
        } yield xs :+ UpdatePortfolioItemInput(sym, rate)
      }
    parsed match {
      case Left(msg)    => Future.failed(BadRequestException(msg))
      case Right(items) =>
        updatePortfolioUsecase.run(UpdatePortfolioUsecaseInput(items)).recoverWith {
          case UpdatePortfolioUsecaseException.InvalidPortfolio(reason) =>
            Future.failed(BadRequestException(reason))
        }
    }
  }
}

File: php/src/CodingInterview/Presentation/PortfolioController.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Presentation;

use Folio\CodingInterview\Application\Usecase\Portfolio\GetLatestPortfolioUsecase;
use Folio\CodingInterview\Application\Usecase\Portfolio\UpdatePortfolioInvalidPortfolioException;
use Folio\CodingInterview\Application\Usecase\Portfolio\UpdatePortfolioItemInput;
use Folio\CodingInterview\Application\Usecase\Portfolio\UpdatePortfolioUsecase;
use Folio\CodingInterview\Application\Usecase\Portfolio\UpdatePortfolioUsecaseInput;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\StockSymbol;

final class PortfolioItemDto
{
    public function __construct(
        public readonly string $symbol,
        public readonly string $rate,
    ) {}
}

final class GetOptimalPortfolioResponse
{
    /** @param PortfolioItemDto[] $portfolios */
    public function __construct(public readonly array $portfolios) {}
}

final class UpdateOptimalPortfolioRequest
{
    /** @param PortfolioItemDto[] $portfolios */
    public function __construct(public readonly array $portfolios) {}
}

final class PortfolioController
{
    public function __construct(
        private readonly GetLatestPortfolioUsecase $getLatestPortfolioUsecase,
        private readonly UpdatePortfolioUsecase $updatePortfolioUsecase,
    ) {}

    public function getOptimalPortfolio(): GetOptimalPortfolioResponse
    {
        $out = $this->getLatestPortfolioUsecase->run();
        $items = [];
        foreach ($out->items as $i) {
            $items[] = new PortfolioItemDto($i->symbol->value, $i->rate->toString());
        }
        return new GetOptimalPortfolioResponse($items);
    }

    public function updateOptimalPortfolio(UpdateOptimalPortfolioRequest $req): void
    {
        $items = [];
        foreach ($req->portfolios as $dto) {
            $sym = StockSymbol::fromStringOrNull($dto->symbol);
            if ($sym === null) {
                throw new BadRequestException("unknown symbol: {$dto->symbol}");
            }
            try {
                $rate = new BigDecimal($dto->rate);
            } catch (\Throwable $e) {
                throw new BadRequestException("invalid rate: {$dto->rate}");
            }
            $items[] = new UpdatePortfolioItemInput($sym, $rate);
        }
        try {
            $this->updatePortfolioUsecase->run(new UpdatePortfolioUsecaseInput($items));
        } catch (UpdatePortfolioInvalidPortfolioException $e) {
            throw new BadRequestException($e->getMessage());
        }
    }
}
